### PR TITLE
Tridiagonal eigensolver (local)

### PR DESCRIPTION
**Pull request by teonnik**
# TODO

- [x] z-vector formation: matrix-vecrtor multiplication of orthogonal matrix (eigenvectors) of the sub-tridiagonal system with Cuppen's vector (i.e. the rank-1 update vector in Cuppen's decomposition)
- [x] Use the recently provided laed4 API from lapackpp for the solution of the secular equation of the rank-1 updated diagonal system (instead of trying to implement your own...)
- [x] Implement deflation for the rank-1 updated diagonal system: DLA-F APIs for Given's rotation
- [x] Matrix-multipy (GEMM) the eigenvectors of the sub-tridiagonal systems with the eigenvectors of the rank-1 updated diagonal system
- [x] Improve and expand unit-testing of various components

Currently the algorithm is not in a working state. Most components are WIP. I will continue pushing changes in this PR as remaining parts are getting ready.

### CODE REVIEW

File: include/dlaf/lapack/tile.h [hunk 1]
```diff
@@ -469,6 +572,11 @@ void potrf(cusolverDnHandle_t handle, const blas::Uplo uplo, const matrix::Tile<
   auto info = potrfInfo(handle, uplo, a);
   assertExtendInfo(dlaf::cusolver::assertInfoHegst, handle, std::move(info));
 }
+
+template <class T>
+void stedc(cusolverDnHandle_t, const Tile<BaseType<T>, Device::CPU>&, const Tile<T, Device::CPU>&) {
+  DLAF_STATIC_UNIMPLEMENTED(sizeof(T) == 0);
```
Comment (msimberg) [inline include/dlaf/lapack/tile.h:578]: I was just browsing and wanted to add a reminder to change this:

```suggestion
  DLAF_STATIC_UNIMPLEMENTED(T);
```

I haven't looked closely at this PR otherwise.

File: include/dlaf/eigensolver/tridiag_solver/permutations.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+//             sz.cols()
+//
+//
+// [1]: The portion of each input column or row is defined by the interval [in_offset, in_offset +
+//      sz.col()) or the interval [in_offset, in_offset + sz.row()) respectively.
+// [2]: The input submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `in_tiles`
+// [3]: The subregion is defined by `begin` and `sz`
+// [4]: The output submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `out_tiles`
+//
+// clang-format off
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/permutations.h:63]: Do we need this?

File: include/dlaf/eigensolver/tridiag_solver/permutations.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    const std::vector<matrix::Tile<T, Device::CPU>>& in_tiles,
+    std::vector<matrix::Tile<T, Device::CPU>>& out_tiles
+    ) {
+  // clang-format on
+
+  // Iterate over rows if `coord == Coord::Row` otherwise iterate over columns
+  SizeType dim1 = sz.cols();
+  SizeType dim2 = sz.rows();
+  if constexpr (coord == Coord::Row)
+    std::swap(dim1, dim2);
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/permutations.h:80]: Does swapping the dimensions like this have any effect on the access pattern into the tile? Not a blocker and can be left as future optimization, but just curious now.

File: test/unit/eigensolver/test_tridiag_solver_permutations.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+using namespace dlaf;
+using namespace dlaf::test;
+using namespace dlaf::eigensolver::internal;
+
+template <typename Type>
+class TridiagEigensolverPermutationsTest : public ::testing::Test {};
+TYPED_TEST_SUITE(TridiagEigensolverPermutationsTest, RealMatrixElementTypes);
+
+// Initializes square input and output matices of size (n x n) and block size (nb x nb). The matrices are
+// discribed by a distribution and an array of input and output tiles respectively. Note that in contrast
```
Comment (msimberg) [inline test/unit/eigensolver/test_tridiag_solver_permutations.cpp:26]:
```suggestion
// described by a distribution and an array of input and output tiles respectively. Note that in contrast
```

File: test/unit/eigensolver/test_tridiag_solver_permutations.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+//
+// The elements of the input matrix are as follows:
+//
+// `in_ij = i + j`
+//
+// The output matrix is set to zero.
+//
+template <class T>
+std::tuple<matrix::Distribution, std::vector<matrix::Tile<T, Device::CPU>>,
+           std::vector<matrix::Tile<T, Device::CPU>>>
```
Comment (msimberg) [inline test/unit/eigensolver/test_tridiag_solver_permutations.cpp:37]: Just curious, is this a type that we would want to make more formal? Would it be possible to use `Matrix` here even though it's overkill?

**Review by msimberg — COMMENTED** [review]
Some initial comments. Looking good to me already. Bonus points for the diagrams :wink:

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+  ColType* c_ptr = coltypes_tiles[0].get().ptr(zero_idx);
+  const SizeType* i_ptr = perm_d_tiles[0].get().ptr(zero_idx);
+
+  return applyDeflationToArrays(rho_fut.get(), tol_fut.get(), n, d_ptr, i_ptr, z_ptr, c_ptr);
+}
+
+// Partitions `p_ptr` based on a `ctype` in `c_ptr` array.
+//
+// Returns the number of elements in the second partition.
+inline SizeType partitionColType(SizeType len, ColType ctype, const ColType* c_ptr, SizeType* i_ptr) {
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:369]: What is a typical value of `len` here?

File: include/dlaf/eigensolver/tridiag_solver/permutations.h (hunk 2 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/permutations.h:63]: It's not really needed, I just put it there because it made it a little bit easier to parse the arguments. I will remove it.

File: test/unit/eigensolver/test_tridiag_solver_permutations.cpp (hunk 5 above)
Comment (teonnik) [reply test/unit/eigensolver/test_tridiag_solver_permutations.cpp:37]: Perhaps we would indeed want to make this a formal type. A pair of `matrx::Distribution` and `std::vector<matrix::Tile<>>` represent a synchronous matrix as opposed to `Matrix<>` which stores futures and is an asynchronous matrix.

To use `Matrix<>`, future of tiles have to be first collected in a container such as `std::vector<future<Tile<>>>` and then unwrapped in a kernel passed to `pika::dataflow()` or `ex::when_all_vector()`.

File: include/dlaf/eigensolver/tridiag_solver/merge.h (hunk 6 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/merge.h]: This can be as large as the number of rows or columns of a distributed matrix. That is, for a big run of a dense 10^6 x 10^6 matrix, len = 10^6.

File: test/unit/eigensolver/test_tridiag_solver_permutations.cpp (hunk 5 above)
Comment (msimberg) [reply test/unit/eigensolver/test_tridiag_solver_permutations.cpp:37]: I don't know how useful this would be in other situations, so to avoid overdesigning this type would it be sufficient to just introduce something like:
```
template <typename T, Device D>
struct SyncMatrix {
  matrix::Distribution dist;
  std::vector<matrix::Tile<T, D>> tiles;
};
```
just to document the type at least a little bit? If it's useful elsewhere in the future we can further formalize it, but I at least dislike plain `tuple`s where a struct could be used instead.

File: include/dlaf/eigensolver/tridiag_solver/mc.h [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                   Matrix<T, Device::CPU>& mat_ev);
+};
+
+/// Splits [i_begin, i_end] in the middle and waits for all splits on [i_begin, i_middle] and [i_middle +
+/// 1, i_end] before saving the triad <i_begin, i_middle, i_end> into `indices`.
+///
+/// The recursive calls span a binary tree which is traversed in depth first left-right-root order. That
+/// is also the order of triads in `indices`.
+///
+/// Note: the intervals are all closed!
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/mc.h:41]: I suppose this also means you will never need empty intervals? If yes, would some assertions about this be in place?

File: include/dlaf/eigensolver/tridiag_solver/mc.h [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+///
+inline void splitIntervalInTheMiddleRecursively(
+    SizeType i_begin, SizeType i_end, std::vector<std::tuple<SizeType, SizeType, SizeType>>& indices) {
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/mc.h:44]: I have the same comment about `std::tuple<SizeType, SizeType, SizeType>` as about the "sync matrix", i.e. it'd be really nice to give names to those `SizeType`s by using a struct instead.

File: include/dlaf/eigensolver/tridiag_solver/mc.h [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+  if (i_begin == i_end)
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/mc.h:45]: I think this answers my other question about empty intervals...

File: include/dlaf/eigensolver/tridiag_solver/mc.h [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+//
+// Substracts the offdiagonal element at the split from the top and bottom diagonal elements and returns
+// the offdiagonal element. The split is between the last row of the top tile and the first row of the
+// bottom tile.
+//
+template <class T>
+T cuppensTileDecomposition(const matrix::Tile<T, Device::CPU>& top,
+                           const matrix::Tile<T, Device::CPU>& bottom) {
+  (void) top;
+  (void) bottom;
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/mc.h:77]:
```suggestion
  silenceUnusedWarningFor(top, bottom);
```
or... why is the `(void)` there in the first place? Did they use to be unused?

File: include/dlaf/eigensolver/tridiag_solver/mc.h [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+std::vector<pika::shared_future<T>> cuppensDecomposition(Matrix<T, Device::CPU>& mat_trd) {
+  SizeType i_end = mat_trd.distribution().nrTiles().rows() - 1;
+  std::vector<pika::shared_future<T>> offdiag_vals;
+  offdiag_vals.reserve(to_sizet(i_end));
+
+  for (SizeType i_split = 0; i_split < i_end; ++i_split) {
+    // Cuppen's tridiagonal decomposition
+    offdiag_vals[to_sizet(i_split)] =
+        pika::dataflow(pika::unwrapping(cuppensTileDecomposition_o), mat_trd(LocalTileIndex(i_split, 0)),
+                       mat_trd(LocalTileIndex(i_split + 1, 0)));
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/mc.h:100]:
```suggestion
        when_all(mat_trd.readwrite_sender(LocalTileIndex(i_split, 0)),
                       mat_trd.readwrite_sender(LocalTileIndex(i_split + 1, 0))) | transform(cuppensTileDecomposition_o) | make_future();
```
modulo namespaces/typos/unclosed parentheses...

File: include/dlaf/eigensolver/tridiag_solver/mc.h [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, float)
+DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, double)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, std::complex<float>)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, std::complex<double>)
+
+#ifdef DLAF_WITH_CUDA
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, float)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, double)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, std::complex<float>)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, std::complex<double>)
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/mc.h:239]: To be enabled in a separate PR or this PR?

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 13]
```diff
@@ -0,0 +1,923 @@
...
+  WorkSpace<T> ws{Matrix<T, Device::CPU>(ev_distr),
+                  Matrix<T, Device::CPU>(ev_distr),
+                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<ColType, Device::CPU>(vec_size, vec_tile_size),
+                  Matrix<ColType, Device::CPU>(vec_size, vec_tile_size)};
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:140]: Designated initializers would be nice to have here :disappointed: Without designated initializers I sometimes find
```
mystruct s;
s.a = some_a;
s.b = some_b;
...
```
nicer because it makes it (more) obvious if you've put a variable in the wrong member variable, but I don't mind the way it is right now either.

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 14]
```diff
@@ -0,0 +1,923 @@
...
+    whenAllLift(tile_row, rho_fut, mat_ev.read_sender(mat_ev_idx), z.readwrite_sender(z_idx)) |
+        copyTileRowAndNormalize(Policy<Backend::MC>(thread_priority::normal)) | start_detached();
+  }
+}
+
+// Multiply by factor 2 to account for the normalization of `z` vector and make sure rho > 0 f
+//
+template <class T>
+pika::future<T> scaleRho(pika::shared_future<T> rho_fut) {
+  return pika::dataflow(pika::unwrapping([](T rho) { return 2 * std::abs(rho); }), std::move(rho_fut));
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:198]: How many of the `dataflow`s in this file do you think you could change to sender adaptors? And for the ones that you can't at the moment, what is blocking you from using sender adaptors?

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 15]
```diff
@@ -0,0 +1,923 @@
...
+    const SizeType* i_ptr = index_tiles[0].get().ptr(zero_idx);
+    // save in variable avoid releasing the tile too soon
+    auto d_tile = d_tiles[0].get();
+    auto z_tile = z_tiles[0].get();
+    auto c_tile = c_tiles[0].get();
+    T* d_ptr = d_tile.ptr(zero_idx);
+    T* z_ptr = z_tile.ptr(zero_idx);
+    ColType* c_ptr = c_tile.ptr(zero_idx);
+
+    return applyDeflationToArrays(rho_fut.get(), tol_fut.get(), n, i_ptr, d_ptr, z_ptr, c_ptr);
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:593]: Could the tiles be passed by reference to `applyDeflationToArrays` instead of passing pointers?

Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/merge.h:593]: Ah, I think I've answered my own question... You do it for the offsets, correct? It would be really nice to have a (sub)view into a tile (or do we already have some helpers for that? @albestro?).

File: include/dlaf/lapack/tile.h [hunk 16]
```diff
@@ -468,6 +578,11 @@ void potrf(cusolverDnHandle_t handle, const blas::Uplo uplo, const matrix::Tile<
   auto info = potrfInfo(handle, uplo, a);
   assertExtendInfo(dlaf::cusolver::assertInfoHegst, handle, std::move(info));
 }
+
+template <class T>
+void stedc(cusolverDnHandle_t, const Tile<BaseType<T>, Device::CPU>&, const Tile<T, Device::CPU>&) {
```
Comment (rasolca) [inline include/dlaf/lapack/tile.h:583]:
```suggestion
void stedc(cusolverDnHandle_t, const Tile<BaseType<T>, Device::CPU>&, const Tile<T, Device::GPU>&) {
```

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 17]
```diff
@@ -0,0 +1,983 @@
...
+};
+
+struct ColTypeLens {
+  SizeType num_uphalf;
+  SizeType num_dense;
+  SizeType num_lowhalf;
+  SizeType num_deflated;
+};
+
+// Auxialiary matrix and vectors used for the D&C algorithm
```
Comment (rasolca) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:69]:
```suggestion
// Auxiliary matrix and vectors used for the D&C algorithm
```

File: include/dlaf/eigensolver/tridiag_solver/mc.h (hunk 7 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/mc.h:43]: By empty intervals, do you mean e.g. `[0, 0]` or `[]`? The former is a valid interval that contains only `0`, the latter is not something I need and will perhaps benefit from an assertion.

I have generally been lax on assertions with internal parts of the code.

Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/mc.h:43]: I meant really empty (`[]`) just because closed intervals make representing those impossible/difficult (unless you allow something like `[i, i-1]`). But if you indeed don't need those then a few assertions might be useful.

File: include/dlaf/eigensolver/tridiag_solver/mc.h (hunk 8 above)
Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/mc.h:46]: Sounds good and fine by me to do that in a separate PR.

Comment (albestro) [reply include/dlaf/eigensolver/tridiag_solver/mc.h:46]: Just another suggestion along these lines for refactoring activity.

It may be also nice to define a type for closed intervals. Currently they are represented as two independent `SizeType`, but having them merged in a struct (possibly with some checks, e.g. `i_begin <= i_end`, and some functions, like `split` perhaps) may improve API (currently it is a lot "C-style") and maintainability.

I don't know yet if they can be extended to be useful also in-place of `std::tuple<SizeType, SizeType, SizeType>` which seems to represent a closed range with a pivot point, but that can be reasoned about too.

**Review by albestro — COMMENTED** [review]
First (partial) round of review.

I'm going to give a look at `merge.h` and tests.

File: include/dlaf/eigensolver/tridiag_solver.h [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+//
+/// @param d [out] (n x 1) local matrix holding the eigenvalues of the the symmetric tridiagonal mat_trd
+//
+/// @param mat_ev [out]    (n x n) local matrix holding the eigenvectors of the the symmetric tridiagonal
+/// matrix on exit.
+///
+/// @pre mat_trd and mat_ev are local matrices
+/// @pre mat_trd has 2 columns
+/// @pre mat_ev is a square matrix
+/// @pre mat_ev has a square block size
```
[the remark this answers is not in the document]
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver.h:35]: No, `mat_trd` is (n x 2). `d` holds the resulting eigenvalues of the tridiagonal system `mat_trd`. It's dimensions are (n x 1).

File: include/dlaf/eigensolver/tridiag_solver/mc.h (hunk 8 above)
Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/mc.h:46]: I suggest we open an issue about this and schedule the cleanup for another PR, so as not to block this further.

File: include/dlaf/eigensolver/tridiag_solver/mc.h (hunk 10 above)
Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/mc.h:79]: What was the answer to this?

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 19]
```diff
@@ -0,0 +1,1018 @@
...
+            T weight = (di == dj) ? evec_el : evec_el / (di - dj);
+            ws_el = (j == 0) ? weight : weight * ws_el;
+          }
+        }
+      };
+      ex::when_all(ex::keep_future(k_fut), diag.read_sender(GlobalTileIndex(i_tile, 0)),
+                   diag.read_sender(GlobalTileIndex(j_tile, 0)),
+                   evecs.read_sender(GlobalTileIndex(i_tile, j_tile)),
+                   ws.readwrite_sender(GlobalTileIndex(i_tile, j_tile))) |
+          di::transform(di::Policy<Backend::MC>(), std::move(mul_rows)) | ex::ensure_started();
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:737]:
```suggestion
      ex::when_all(ex::keep_future(k_fut), diag.read_sender(GlobalTileIndex(i_tile, 0)),
                   diag.read_sender(GlobalTileIndex(j_tile, 0)),
                   evecs.read_sender(GlobalTileIndex(i_tile, j_tile)),
                   ws.readwrite_sender(GlobalTileIndex(i_tile, j_tile))) |
          di::transformDetached(di::Policy<Backend::MC>(), std::move(mul_rows));
```

and same elsewhere. Alternatively:

```suggestion
      ex::start_detached(ex::when_all(ex::keep_future(k_fut), diag.read_sender(GlobalTileIndex(i_tile, 0)),
                   diag.read_sender(GlobalTileIndex(j_tile, 0)),
                   evecs.read_sender(GlobalTileIndex(i_tile, j_tile)),
                   ws.readwrite_sender(GlobalTileIndex(i_tile, j_tile))) |
          di::transform(di::Policy<Backend::MC>(), std::move(mul_rows)));
```

These are functionally equivalent to what you have, but semantically clearer since you're not using the sender returned from `ensure_started`.

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 20]
```diff
@@ -0,0 +1,1018 @@
...
+void copySubMatrix(SizeType i_begin, SizeType i_end, Matrix<const T, Source>& source,
+                   Matrix<T, Destination>& dest) {
+  namespace ex = pika::execution::experimental;
+
+  for (SizeType j = i_begin; j <= i_end; ++j) {
+    for (SizeType i = i_begin; i <= i_end; ++i) {
+      ex::when_all(source.read_sender(LocalTileIndex(i, j)),
+                   dest.readwrite_sender(LocalTileIndex(i, j))) |
+          matrix::copy(dlaf::internal::Policy<matrix::internal::CopyBackend_v<Source, Destination>>{}) |
+          ex::start_detached();
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:887]:
```suggestion
      ex::start_detached(ex::when_all(source.read_sender(LocalTileIndex(i, j)),
                   dest.readwrite_sender(LocalTileIndex(i, j))) |
          matrix::copy(dlaf::internal::Policy<matrix::internal::CopyBackend_v<Source, Destination>>{}));
```
because the reference implementation does not allow piping into `start_detached`.

File: include/dlaf/permutations/general/impl.h [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+  namespace ex = pika::execution::experimental;
+
+  const matrix::Distribution& distr = mat_in.distribution();
+  TileElementSize sz_last_tile = distr.tileSize(GlobalTileIndex(i_end, i_end));
+  SizeType m = distr.globalTileElementDistance<Coord::Row>(i_begin, i_end) + sz_last_tile.rows();
+  SizeType n = distr.globalTileElementDistance<Coord::Col>(i_begin, i_end) + sz_last_tile.cols();
+  matrix::Distribution subm_distr(LocalElementSize(m, n), distr.blockSize());
+  SizeType ntiles = i_end - i_begin + 1;
+
+  // Note: there is an issue compiling with dlaf::internal::transform()
```
Comment (msimberg) [inline include/dlaf/permutations/general/impl.h:165]: Did `transform<false>` not help here or was there another problem?

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 22]
```diff
@@ -0,0 +1,1018 @@
...
+
+// Multiply by factor 2 to account for the normalization of `z` vector and make sure rho > 0 f
+//
+template <class T>
+pika::future<T> scaleRho(pika::shared_future<T> rho_fut) {
+  // Note: `keep_future()` is needed here even though `T` is a scalar type (not a `Tile<>`) otherwise there
+  // is a compile-time error. A fix will be available in pika@0.6.0 : https://github.com/pika-org/pika/pull/282
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
+  return ex::keep_future(std::move(rho_fut)) |
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:236]: This should work without `keep_future`. Did you find that you needed it to make things compile? If yes, that may indicate a bug. A scalar `T` should definitely be copiable from a `const&`.

File: include/dlaf/eigensolver/tridiag_solver/mc.h [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+std::vector<pika::shared_future<T>> cuppensDecomposition(Matrix<T, Device::CPU>& mat_trd) {
+  SizeType i_end = mat_trd.distribution().nrTiles().rows() - 1;
+  std::vector<pika::shared_future<T>> offdiag_vals;
+  offdiag_vals.reserve(to_sizet(i_end));
+
+  for (SizeType i_split = 0; i_split < i_end; ++i_split) {
+    // Cuppen's tridiagonal decomposition
+    offdiag_vals.push_back(pika::dataflow(pika::unwrapping(cuppensTileDecomposition_o),
+                                          mat_trd(LocalTileIndex(i_split, 0)),
+                                          mat_trd(LocalTileIndex(i_split + 1, 0))));
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/mc.h:104]:
```suggestion
    offdiag_vals.push_back(ex::when_all(mat_trd(LocalTileIndex(i_split, 0)),
                                        mat_trd(LocalTileIndex(i_split + 1, 0))) |
                           transform(Policy<MC>(), cuppensTileDecomposition_o) | ex::make_future()));
```
probably with some parentheses fixed... Even better would be to not make a future at all and use `split`, but I'm happy to change that at a later point as well.

**Comment by msimberg** [comment]
Thanks @teonnik for making the `dataflow` to sender changes. The only thing I would insist on in this PR is the `ensure_started` to `start_detached` change. For the rest of the open comments I would ask for a response/issue to keep track of it.

File: include/dlaf/permutations/general/impl.h (hunk 21 above)
Comment (teonnik) [reply include/dlaf/permutations/general/impl.h]: I hadn't merged the teo/permutations branch here. This is a stale comment.

File: include/dlaf/eigensolver/tridiag_solver/merge.h (hunk 19 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/merge.h]: https://github.com/eth-cscs/DLA-Future/pull/436/commits/966bd5e36d84d2b231bbaab91fcaaae43140d0a1

File: include/dlaf/eigensolver/tridiag_solver/mc.h (hunk 23 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/mc.h]: https://github.com/eth-cscs/DLA-Future/pull/436/commits/f9e605a3115a9aa588bd3b099b0b4f0ee6d9c607

Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/mc.h]: Thanks for updating! Issue for removing the `make_future`: https://github.com/eth-cscs/DLA-Future/issues/588.

File: include/dlaf/eigensolver/tridiag_solver/merge.h (hunk 22 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/merge.h:236]: Locally it does compile for me without `keep_future()` but that is because I built `pika@0.5.0` with a patch containing a part of https://github.com/pika-org/pika/pull/282 (merged for `pika@0.6.0`). The error I reported to you a few days ago on slack was this : https://pastebin.com/uaGzmsuw , then you suggested to check if https://github.com/pika-org/pika/pull/282 will help and it did.

Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/merge.h:236]: Ah, sorry... it's even in the comment right above :facepalm: I remember now what this was about. Opened https://github.com/eth-cscs/DLA-Future/issues/590 for my own sake.

File: include/dlaf/eigensolver/tridiag_solver/mc.h (hunk 12 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/mc.h:245]: This will be enabled in a separate PR.

File: include/dlaf/eigensolver/tridiag_solver/merge.h (hunk 13 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/merge.h]: I agree this is not ideal but I suggest we leave it for a future refactoring.

File: test/unit/eigensolver/test_tridiag_solver.cpp [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+  solveLaplace1D<TypeParam>(16, 8);
+}
+
+TYPED_TEST(TridiagEigensolverTest, Laplace1D_n16_nb4) {
+  solveLaplace1D<TypeParam>(16, 4);
+}
+
+TYPED_TEST(TridiagEigensolverTest, Laplace1D_n16_nb5) {
+  solveLaplace1D<TypeParam>(16, 5);
+}
```
Comment (rasolca) [inline test/unit/eigensolver/test_tridiag_solver.cpp:151]: I would consider adding a test with random elements as well.

Comment (teonnik) [reply test/unit/eigensolver/test_tridiag_solver.cpp:249]: https://github.com/eth-cscs/DLA-Future/pull/436/commits/b91ef2f7d8c5c3fbaed1a311d8dc61046a8487db

**Comment by teonnik** [comment]
bors try

**Comment by teonnik** [comment]
bors try

File: test/unit/eigensolver/test_tridiag_solver.cpp [hunk 25]
```diff
@@ -76,36 +78,38 @@ TYPED_TEST(TridiagEigensolverTest, CuppensDecomposition) {
...
   matrix::Matrix<T, Device::CPU> evecs(LocalElementSize(n, n), TileElementSize(nb, nb));
 
   // Tridiagonal matrix : 1D Laplacian
   auto mat_trd_fn = [](GlobalElementIndex el) {
     if (el.col() == 0)
       // diagonal
-      return T(2);
+      return RealParam(2);
     else
       // off-diagoanl
```
Comment (rasolca) [inline test/unit/eigensolver/test_tridiag_solver.cpp:95]:
```suggestion
      // off-diagonal
```

**Review by rasolca — APPROVED** [review]
Looks good. Just a typo to be fixed.

**Comment by teonnik** [comment]
bors try

**Comment by teonnik** [comment]
bors try

**Comment by rasolca** [comment]
bors try

**Comment by teonnik** [comment]
bors try

**Comment by rasolca** [comment]
There is still a problem with the ROCM build, that will be addressed in another PR (See #614)
